ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplLoadTask.java

```diff
@@ -279,6 +292,72 @@ a database ( directory )
   }
 
+  /**
+   * Cleanup/drop tables from the given database which are bootstrapped by input dump dir.
+   * @throws HiveException Failed to drop the tables.
+   * @throws IOException File operations failure.
+   * @throws InvalidInputException Invalid input dump directory.
+   */
+  private void bootstrapRollbackTask() throws HiveException, IOException, InvalidInputException {
+    Path bootstrapDirectory = new PathBuilder(work.bootstrapDumpToRollback)
+            .addDescendant(ReplUtils.INC_BOOTSTRAP_ROOT_DIR_NAME).build();
+    FileSystem fs = bootstrapDirectory.getFileSystem(conf);
+
+    if (!fs.exists(bootstrapDirectory)) {
+      throw new InvalidInputException("Input bootstrap dump directory to rollback doesn't exist: "
+              + bootstrapDirectory);
+    }
+
+    FileStatus[] fileStatuses = fs.listStatus(bootstrapDirectory, EximUtil.getDirectoryFilter(fs));
+    if ((fileStatuses == null) || (fileStatuses.length == 0)) {
+      throw new InvalidInputException("Input bootstrap dump directory to rollback is empty: "
+              + bootstrapDirectory);
+    }
+
+    if (StringUtils.isNotBlank(work.dbNameToLoadIn) && (fileStatuses.length > 1)) {
+      throw new InvalidInputException("Multiple DB dirs in the dump: " + bootstrapDirectory
+                      + " is not allowed to load to single target DB: " + work.dbNameToLoadIn);
+    }
+
+    for (FileStatus dbDir : fileStatuses) {
+      Path dbLevelPath = dbDir.getPath();
+      String dbNameInDump = dbLevelPath.getName();
+
+      List<String> tableNames = new ArrayList<>();
+      RemoteIterator<LocatedFileStatus> filesIterator = fs.listFiles(dbLevelPath, true);
+      while (filesIterator.hasNext()) {
+        Path nextFile = filesIterator.next().getPath();
+        String filePath = nextFile.toString();
+        if (filePath.endsWith(EximUtil.METADATA_NAME)) {
+          // Remove dbLevelPath from the current path to check if this _metadata file is under DB or
+          // table level directory.
+          String replacedString = filePath.replace(dbLevelPath.toString(), "");
+          if (!replacedString.equalsIgnoreCase(EximUtil.METADATA_NAME)) {
+            tableNames.add(nextFile.getParent().getName());
+          }
+        }
+      }
+
+      // No tables listed in the DB level directory to be dropped.
+      if (tableNames.isEmpty()) {
+        LOG.info("No tables are listed to be dropped for Database: {} in bootstrap dump: {}",
+                dbNameInDump, bootstrapDirectory);
+        continue;
+      }
+
+      // Drop all tables bootstrapped from previous dump.
+      // Get the target DB in which previously bootstrapped tables to be dropped. If user specified
+      // DB name as input in REPL LOAD command, then use it.
+      String dbName = (StringUtils.isNotBlank(work.dbNameToLoadIn) ? work.dbNameToLoadIn : dbNameInDump);
+
+      Hive db = getHive();
+      for (String table : tableNames) {
+        db.dropTable(dbName + "." + table, true);
```

Comment:
What happens to the underlying external table directory when an external table is dropped? Consider the case in the test where an external table is dropped and a managed table with the same name is created. Following sequence of events would leave a dangling external table directory in the file system
1.  an external table is created
2. an incremental bootstrap dump is taken
3. external table is dropped and a managed table with the same name is created
4. previous incremental bootstrap dump fails to load after it has created the external table directory and copied files
5. a new incremental bootstrap dump is taken and loaded with location of the previous incremental bootstrap dump specified.
6. new incremental bootstrap dump is loaded

Reply:
Dropping an external table doesn't delete the directory. Also, managed table doesn't use the same location as external tables and so there won't be any conflict.
So, re-bootstrap, drop the old external tables and incremental will create the new managed table and the operation will be successful. Probably, I will add a test to cover this.

Reply:
Well, I think my current test case covers this case. So, it's done.

Reply:
The external table directory will still be left behind. That's the problem I am trying to highlight here.

Reply:
That's not a problem. It is expected behaviour for external tables where Hive shouldn't delete those directories. Even in source these directories will remain there but the table would be dropped. Also, if the same table is found  in new bootstrap dump, then it will avoid re-copying those file again.

Reply:
Ok. Thanks.